src/components/connectAdvanced.js

```diff
@@ -100,203 +90,169 @@ export default function connectAdvanced(
 
     const displayName = getDisplayName(wrappedComponentName)
 
+    let PureWrapper
+
+    if (forwardRef) {
+      class PureWrapperRef extends Component {
+        shouldComponentUpdate(nextProps) {
+          return nextProps.derivedProps !== this.props.derivedProps
+        }
+
+        render() {
+          let { forwardRef, derivedProps } = this.props
+          return <WrappedComponent {...derivedProps} ref={forwardRef} />
+        }
+      }
+      PureWrapperRef.propTypes = {
+        derivedProps: propTypes.object,
+        forwardRef: propTypes.oneOfType([
+          propTypes.func,
+          propTypes.object
+        ])
+      }
+      PureWrapper = PureWrapperRef
+    } else {
+      class PureWrapperNoRef extends Component {
+        shouldComponentUpdate(nextProps) {
+          return nextProps.derivedProps !== this.props.derivedProps
+        }
+
+        render() {
+          return <WrappedComponent {...this.props.derivedProps} />
+        }
+      }
+      PureWrapperNoRef.propTypes = {
+        derivedProps: propTypes.object,
+      }
+      PureWrapper = PureWrapperNoRef
+    }
+
     const selectorFactoryOptions = {
       ...connectOptions,
       getDisplayName,
       methodName,
       renderCountProp,
       shouldHandleStateChanges,
       storeKey,
-      withRef,
       displayName,
       wrappedComponentName,
       WrappedComponent
     }
 
-    // TODO Actually fix our use of componentWillReceiveProps
-    /* eslint-disable react/no-deprecated */
-
-    class Connect extends Component {
-      constructor(props, context) {
-        super(props, context)
-
-        this.version = version
-        this.state = {}
-        this.renderCount = 0
-        this.store = props[storeKey] || context[storeKey]
-        this.propsMode = Boolean(props[storeKey])
-        this.setWrappedInstance = this.setWrappedInstance.bind(this)
-
-        invariant(this.store,
-          `Could not find "${storeKey}" in either the context or props of ` +
-          `"${displayName}". Either wrap the root component in a <Provider>, ` +
-          `or explicitly pass "${storeKey}" as a prop to "${displayName}".`
+    const OuterBase = connectOptions.pure ? PureComponent : Component
+
+    class Connect extends OuterBase {
+      constructor(props) {
+        super(props)
+        invariant(forwardRef ? !props.props[storeKey] : !props[storeKey],
+          'Passing redux store in props has been removed and does not do anything. ' +
+          'To use a custom redux store for a single component, ' +
+          'create a custom React context with React.createContext() and pass the Provider to react-redux\'s provider ' +
```

Comment:
Error message will need to be updated appropriately.

Also, can we do this _only_ in dev?  Just want to minimize the startup path in prod.